tsl/src/compression/algorithms/uuid_compress.c

```diff
@@ -0,0 +1,594 @@
+#define HLL_ERROR_RATE 0.065
+#define HLL_MIN_CARDINALITY 20
+
+/*
+ * The UUID compressor is using delta delta compression for the first 8
+ * bytes of the UUID and stores the rest as a uint64_vec. At the same time
+ * it keeps track of the cardinality of values. if the cardinality
+ * indicates that we are better off with the dictionary compressor, we will
+ * recompress it at the end.
+ */
+typedef struct UuidCompressor
+{
+	/* Delta-delta encoding for timestamp, version and rand_a. */
+	DeltaDeltaCompressor *timestamp;
+
+	/* We store the rand_b and variant parts together as a uint64_vec
+	 * to avoid having to store two separate bitmaps.
+	 */
+	uint64_vec rand_b_and_variant;
+
+	/* HLL state to estimate the cardinality. This is used to check if
+	 * we are better off with recompressing the data as a dictionary.
+	 */
+	hyperLogLogState cardinality;
+
+	/* Number of nulls in the data. */
+	uint16 num_nulls;
+
+	/* Number of elements in the timestamp part. */
+	uint16 num_values;
+} UuidCompressor;
+
+typedef struct ExtendedCompressor
+{
+	Compressor base;
+	UuidCompressor *internal;
+} ExtendedCompressor;
+
+/*
+ * Local helpers
+ */
+static void uuid_compressor_append_uuid(Compressor *compressor, Datum val);
+static void uuid_compressor_append_null_value(Compressor *compressor);
+static void *uuid_compressor_finish_and_reset(Compressor *compressor);
+static void decompression_iterator_init(UuidDecompressionIterator *iter, void *compressed,
+										Oid element_type, bool forward);
+
+const Compressor uuid_compressor_initializer = {
+	.append_val = uuid_compressor_append_uuid,
+	.append_null = uuid_compressor_append_null_value,
+	.is_full = NULL,
+	.finish = uuid_compressor_finish_and_reset,
+};
+
+/*
+ * Compressor framework functions and definitions for the uuid_compress algorithm.
+ */
+
+extern UuidCompressor *
+uuid_compressor_alloc(void)
+{
+	UuidCompressor *compressor = palloc0(sizeof(*compressor));
+	compressor->timestamp = delta_delta_compressor_alloc();
+	uint64_vec_init(&compressor->rand_b_and_variant,
+					CurrentMemoryContext,
+					TARGET_COMPRESSED_BATCH_SIZE);
+	initHyperLogLog(&compressor->cardinality, HLL_BIT_WIDTH);
+	return compressor;
+}
+
+extern void
+uuid_compressor_append_null(UuidCompressor *compressor)
+{
+	delta_delta_compressor_append_null(compressor->timestamp);
+	compressor->num_nulls++;
+}
+
+extern void
+uuid_compressor_append_value(UuidCompressor *compressor, pg_uuid_t next_val)
+{
+	uint64_t components[2];
+	memcpy(components, next_val.data, sizeof(components));
+
+	/* The first component is the timestamp, version and rand_a. */
+	uint64_t timestamp = pg_ntoh64(components[0]);
+	/* The second part is the rand_b and variant. */
+	uint64_t rand_b_and_variant = components[1];
+
+	delta_delta_compressor_append_value(compressor->timestamp, timestamp);
+	uint64_vec_append(&compressor->rand_b_and_variant, rand_b_and_variant);
+
+	uint32 h = hash_bytes((unsigned char *) &next_val, sizeof(next_val));
```

Comment:
I suggest something like `bloom1_hash_16`, should be much faster and have better characteristics than the Postgres functions.